Minimize use of global write lock in typical event loops

bendavid opened the pull request:
This avoids the remaining uses of the global write lock in typical RDataFrame event loops.

There are two parts to this

1) A bug fix in the hashing of streamer info when opening files.  (The hash was including the TKey timestamp and therefore ended up being unique even for files with identical streamer info).

2) Since TTreeProcessorMT can carefully manage its TChains and corresponding TFile and TTree objects, global lists/cleanup and thread safety features can be bypassed or disabled.

Together these changes are enough to almost completely eliminate the use of the global write lock in typical event loops.  The only remaining using is the calls from ```TFile``` to ```TProcessUUID::AddUUID``` and ```TProcessUUID::RemoveUUID```  There are still several remaining places where the global read lock is taken, and these still have some cost in light of the remaining global write lock usage.


These changes have the biggest impact when analyzing a large number of files.  A test case is below.

Produce the test dataset:
```cpp
#include "TFile.h"
#include "TTree.h"
#include "TString.h"
#include <thread>

void testwrite() {

  const unsigned int nfiles = 4000;
  const unsigned int nentries = 1000*1000;

  float outval = 1.;

  for (unsigned int ifile = 0; ifile < nfiles; ++ifile) {
    TFile *fout = TFile::Open(TString::Format("test_%i.root", ifile), "RECREATE");
    TTree *tree = new TTree("tree", "");
    tree->Branch("outval", &outval);
    for (unsigned int ientry = 0; ientry < nentries; ++ientry) {
      tree->Fill();
    }
    tree->Write();
    fout->Close();

    // make sure that each key has a distinct timestamp to maximally
    // provoke different hashes for streamer info in each file
    std::this_thread::sleep_for(1100ms);
  }

}
```

Test event loop:
```python
import ROOT
ROOT.gInterpreter.ProcessLine(".O3")
ROOT.ROOT.EnableImplicitMT()


chain = ROOT.TChain("tree")
chain.Add("test_*.root")

d = ROOT.ROOT.RDataFrame(chain)
res = d.Sum("outval")

resval = res.GetValue()
print(resval)
```

With 256 threads:

+ Baseline:
        Percent of CPU this job got: 391%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 5:17.52

+ +hashing fix
        Percent of CPU this job got: 453%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 3:15.25

+ +hashing fix and reduction of remaining global write locks (this PR):
        Percent of CPU this job got: 1639%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 0:38.98

+ +hashing fix and reduction of remaining global write locks + remove TFile UUID registration (not in this PR anymore):
        Percent of CPU this job got: 19861%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 0:17.52

linev commented:
@phsft-bot build

eguiraud commented:
Super cool, thanks a lot @bendavid ! This is a boon for the case of many (small) files and many cores. The changes to TTreeProcessorMT look good to me, we need Axel or Philippe to take a look at the rest.

Axel-Naumann commented on core/cont/inc/TCollection.h:
```diff
@@ -205,7 +205,7 @@ class TCollection : public TObject {
    void               SetName(const char *name) { fName = name; }
    virtual void       SetOwner(Bool_t enable = kTRUE);
-   virtual bool       UseRWLock();
+   virtual bool UseRWLock(Bool_t enable = true);
```
Could you keep the original indent? (multiple occurrences)

Axel-Naumann commented on io/io/inc/TFile.h:
```diff
@@ -113,6 +113,8 @@ class TFile : public TDirectoryFile {
    TList           *fInfoCache{nullptr};      ///<!Cached list of the streamer infos in this file
    TList           *fOpenPhases{nullptr};     ///<!Time info about open phases
 
+   bool fLocallyManaged = false; ///<! if true, bypass use of global lists
```
Please adjust indent of member name

Axel-Naumann replied:
"Locally" could also mean "not remote"... Maybe a better name could be `fAvoidGlobalRegistration`?

Axel-Naumann commented on tree/tree/src/TChain.cxx:
```diff
@@ -68,38 +68,34 @@ ClassImp(TChain);
-   if (gDirectory) gDirectory->Remove(this);
-   gROOT->GetListOfSpecials()->Add(this);
+   if (!fLocallyManaged) {
+      if (gDirectory)
+         gDirectory->Remove(this);
```
Should `gDirectory` still memory manage `*this` if `fLocallyManaged`?

bendavid replied:
Probably not.  But actually I missed skipping the ```Append``` side of this, which occurs in the TTree constructor (and is already marked as a FIXME: https://github.com/root-project/root/blob/04375696c2e00be2c8881636e9ac9d43d7d55e54/tree/tree/src/TTree.cxx#L899-L903)

I would propose to skip both the ```Append``` and ```Remove``` if ```fLocallyManaged```

pcanal replied:
Actually, I think the `gDirectory->Remove(this);` is now unnecessary (as `TTree::TTree()` no longer add it).

pcanal replied:
And note:t that the code in TTree.cxx:903 should be reacheable from the TChain constructor.

bendavid replied:
You mean should *not* be reachable from the TChain constructor right?  (because it calls the TTree constructor with ```dir = nullptr```)

pcanal replied:
If I am not mistaken, `TChain`'s constructor use the default `TTree` constructor which does 'nothing'.

bendavid replied:
The second ```TChain``` constructor (also relevant here) calls the non-default ```TTree``` constructor.  (But as said the ```gDirectory->Append``` call is anyways skipped because it's called with ```dir = nullptr```, so we can still skip the ```gDirectory->Remove``` call here.)

Axel-Naumann commented on tree/tree/inc/TChain.h:
```diff
@@ -66,8 +67,8 @@ class TChain : public TTree {
 public:
-   TChain();
-   TChain(const char* name, const char* title = "");
+   TChain(bool locallyManaged = false);
+   TChain(const char *name, const char *title = "", bool locallyManaged = false);
```
Can these take an enum class instead of a bool, to make it less likely that they are picked up by users by mistake?

Axel-Naumann reviewed (commented):
Needs review by @pcanal

pcanal commented on io/io/src/TFile.cxx:
```diff
@@ -823,7 +832,7 @@ void TFile::Init(Bool_t create)
    }
 
-   {
+   if (!fLocallyManaged) {
```
Removing:
```
   gROOT->GetUUIDs()->AddUUID(fUUID,this);
```
induces a non-ownership functional change.  Not registered the file UUID will make resolving `TRef` within the file non-functional.

bendavid replied:
Ok, noted.  I think the use of the global lock can probably be avoided in TProcessUUID in favor of a local one.  Could we safely skip the Add and remove of the UUID for the case where ```fNProcessIDs == 0```?

pcanal replied:
By definition no, I think. The add should increment fNProcessIDs.  

>  I think the use of the global lock can probably be avoided in TProcessUUID in favor of a local one. 

Great (by watch out for hidden/indirect risk of dead-lock: if the section with the local lock has any possibility of taking the global lock (indirectly, for example via a call to Print or a GetClass) then there is a risk of dead-lock (risk == guarantee that it will happen somewhere somehow).

bendavid replied:
What I meant was that if fNProcessIDs for the file is 0 (before ```AddUUID```) doesn't this mean that there were no TRefs stored in this file?

bendavid replied:
In fact I'm still not clear on the issue exactly.  Wouldn't skipping the UUID registration only affect TRefs which point directly to the TFile object itself?

pcanal replied:
Sorry I was too quick to answer.  I think you are right.  Skipping the TFile's UUID registration would only affect the TRefs that are pointing to the TFile.  Furthermore, if `TFile::fNProcessIDs` is zero there should not be any `TRef` inside the file ... However `fNProcessIDs` is currently set after the UUID is registered ... but it should be fine to move it before the UUID registration.

bendavid replied:
So actually I'm not sure the ```TFile::fNProcessIDs``` check is relevant in that case, since writing a TRef which points to an object with a UUID does not trigger the writing of the TProcessID.

Also the kIsReferenced bit for the TFile object doesn't seem to be preserved when a file is written and then opened again, so I'm not sure this case can be easily detected.

For the case of a TRef stored in a file and pointing to the same file we could probably handle the needed registration in the TRef streamer while reading.  This wouldn't handle the case of a TRef pointing to a file which is read from a different file or assigned during the  same process though...

pcanal replied:
> This wouldn't handle the case of a TRef pointing to a file which is read from a different file or assigned during the same process though...

That's (one of) the real problem(s) ...  So I can't see how to avoid having to register the UUID ... and we probably need to find ways to make the adding more thread friendly.

bendavid commented:
Rebased and addressed review comments.  The registration and cleanup of the UUID is kept for now because it cannot be safely removed in all cases.  To be addressed in a future PR.

bendavid commented:
(clang-format failures should be from preserving indentation of new data members/functions)

pcanal commented on io/io/src/TFile.cxx:
```diff
@@ -383,6 +383,14 @@ TFile::TFile(const char *fname1, Option_t *option, const char *ftitle, Int_t com
    if (fOption == "NEW")
       fOption = "CREATE";
 
+   if (fOption == "AVOIDGLOBALREGISTRATION") {
```
```suggestion
   if (fOption == "SKIPGLOBALREGISTRATION") {
```

bendavid replied:
This name was suggested by @Axel-Naumann earlier in the discussion.  I have no strong preference.

bendavid replied:
Ok, changed names and swapped logic as suggested.

pcanal reviewed (approved):
Thanks.

bendavid commented:
I had the impression that the test failures were unrelated, but maybe worth running them again...

pcanal commented:
@phsft-bot build

Axel-Naumann commented:
@bendavid you, too, should have the powers to tell phsft-bot to build. Enjoy! :-)

pcanal commented:
humm .. that weird .... I  don't see why the failure on  'ROOT-performance-centos8-multicore/default.' should appear because of this PR but it seems to be ...

pcanal commented:
@Axel-Naumann the build seems 'stuck' on an another platform.  Do you know why?

Axel-Naumann commented:
https://lcgapp-services.cern.ch/root-jenkins/job/root-pullrequests-handler/20430/flowGraphTable/ shows it's two platforms:
- ubuntu 20.04, known, and fixed (the builds are catching up)
- mac11: we had 1/2 the nodes (understood and fixed); builds are catching up.

pcanal commented:
@phsft-bot build just on ROOT-performance-centos8-multicore/default

pcanal commented:
@phsft-bot build just on ROOT-performance-centos8-multicore/default

pcanal commented:
@phsft-bot build just on ROOT-performance-centos8-multicore/default

pcanal commented:
@phsft-bot build just on ROOT-performance-centos8-multicore/default

pcanal commented:
@phsft-bot build with flags -DCTEST_TEST_EXCLUDE_NONE=On

pcanal commented:
The test failure in [fit/combinedFit.py](https://lcgapp-services.cern.ch/root-jenkins/job/root-pullrequests-build/135926/testReport/projectroot/runtutorials/tutorial_fit_combinedFit_py/) is unrelated see https://github.com/root-project/root/pull/3760#issuecomment-1026103150.

pcanal commented:
@bendavid Thank you very much!

pcanal commented:
@phsft-bot build just on ROOT-performance-centos8-multicore/default  with flags -DLLVM_BUILD_TYPE=Debug -DBUILD_TYPE=Debug

guitargeek commented:
This PR was backported to v626 in this PR:
https://github.com/root-project/root/pull/10215

guitargeek commented:
This PR appears to have broken the ouptut of more than one HistFactory workspaces in the same script/executable:
https://github.com/root-project/root/issues/12783

pcanal commented:
Can you check whether revert this commit https://github.com/root-project/root/pull/9486/commits/905d8215e4628a9ae5deae8bb64cf61cfa2a7249 improves the situation?

guitargeek commented:
Unfortunately not. On the contrary, when I revert only that commit, both of the workspaces that HistFactory writes out in my reproducer have this problem of corruption and reduced size, not only the first one.

Probably, this commit is connected to the others in a non-trivial way, you think it's worth/possible to revert all of them in my local ROOT and maybe it works then?

pcanal commented:
My guess is that reverting more of this PR will not change the (worsened) behavior.  However, I could be wrong so it is worth a try.
One thing to compare and check is the content of the `StreamerInfo` record.  To do so for each of the files (and store in a file for comparison):
```
root.exe -b -l -q filename.root -e 'gFile->ShowStreamerInfo();'
```

guitargeek commented:
Okay, thanks then maybe I'll try reverting the full PR tomorrow.

For the streamer info, I have here the ones of the first valid workspace and the second broken one:
1. [streamer_info_1.txt](https://github.com/root-project/root/files/11417549/streamer_info_1.txt)
2. [streamer_info_2.txt](https://github.com/root-project/root/files/11417550/streamer_info_2.txt)
The difference is that a lot is missing in the second one. For some reason, only the classes in the nested `RooStats::HistFactory` namespace survive. There is not even the `RooWorkspace` mentioned, even if it's there according to `gFile->ls()`:
```
Attaching file ws2/WS_combined_ws2_model.root as _file0...
(TFile *) 0x55db6b172f20
TFile**		ws2/WS_combined_ws2_model.root	
 TFile*		ws2/WS_combined_ws2_model.root	
  KEY: RooWorkspace	combined;1	combined
  KEY: TProcessID	ProcessID0;1	a7000b6c-ed54-11ed-94cf-51bb8d80beef
  KEY: TDirectoryFile	SR_hists;1	SR_hists
  KEY: RooStats::HistFactory::Measurement	ws2;1	ws2
```

Does the fact that the infos are missing besides for the classes in `RooStats::HistFactory` help to understand the problem?

pcanal commented:
> Each class that is stored in a files should have a streamer info, right?

Yes.

> The difference is that a lot is missing in the second one. 

This indicates that the problem is in the writing of the file.

pcanal commented:
Continuing in the issue report rather than here.  

https://github.com/root-project/root/issues/12783.
